cdap-ui/app/cdap/components/Replicator/Create/Content/SelectColumns/index.tsx

```diff
@@ -0,0 +1,302 @@
+
+        '& .grid-row': {
+          gridTemplateColumns: '55px 40px 1fr 200px 55px 100px',
+          alignItems: 'center',
+        },
+
+        '& > div[class^="grid-"] .grid-row > div': {
+          paddingTop: 0,
+          paddingBottom: 0,
+        },
+      },
+    },
+    nullableCheckbox: {
+      paddingLeft: 0,
+    },
+    loadingContainer: {
+      textAlign: 'center',
+      marginTop: '100px',
+    },
+  };
+};
+
+interface ISelectColumnsProps extends WithStyles<typeof styles> {
+  tableInfo?: {
+    table: string;
+    database: string;
+  };
+  onSave: (tableKey, columns) => void;
+  initialSelected: List<Map<string, string>>;
+  toggle: () => void;
+  draftId: string;
+}
+
+interface IColumn {
+  name: string;
+  type: string;
+  nullable: boolean;
+}
+
+interface ISelectColumnsState {
+  columns: IColumn[];
+  primaryKeys: string[];
+  selectedColumns: Map<string, Map<string, string>>;
+  loading: boolean;
+  error: any;
+}
+
+class SelectColumnsView extends React.PureComponent<ISelectColumnsProps, ISelectColumnsState> {
+  public state = {
+    columns: [],
+    primaryKeys: [],
+    selectedColumns: Map<string, Map<string, string>>(),
+    loading: true,
+    error: null,
+  };
+
+  public componentDidMount() {
+    this.fetchColumns();
+  }
+
+  private fetchColumns = () => {
+    this.setState({
+      loading: true,
+    });
+
+    const params = {
+      namespace: getCurrentNamespace(),
+      draftId: this.props.draftId,
+    };
+
+    const body = {
+      table: this.props.tableInfo.table,
+      database: this.props.tableInfo.database,
+    };
+
+    MyReplicatorApi.getTableInfo(params, body).subscribe(
+      (res) => {
+        const selectedColumns = {};
+        if (this.props.initialSelected && this.props.initialSelected.size > 0) {
+          this.props.initialSelected.forEach((row) => {
+            selectedColumns[row.get('name')] = row;
+          });
+        }
+
+        this.setState({
+          columns: res.columns,
+          primaryKeys: res.primaryKey,
+          selectedColumns: Map(selectedColumns),
+        });
+      },
+      (err) => {
+        this.setState({ error: err });
+      },
+      () => {
+        this.setState({
+          loading: false,
+        });
+      }
+    );
+  };
+
+  private handleSave = () => {
+    const selectedList = this.state.selectedColumns.toList();
+    const tableKey = generateTableKey(this.props.tableInfo);
+
+    this.props.onSave(tableKey, selectedList);
+    this.props.toggle();
+  };
+
+  private toggleSelected = (row) => {
+    const key = row.name;
+    if (this.state.selectedColumns.get(key)) {
+      this.setState({
+        selectedColumns: this.state.selectedColumns.delete(key),
+      });
+      return;
+    }
+
+    this.setState({
+      selectedColumns: this.state.selectedColumns.set(key, Map({ name: row.name, type: row.type })),
+    });
+  };
+
+  private toggleSelectAll = () => {
+    if (this.state.selectedColumns.size > 0) {
```

Comment:
Does this mean that if they already have some columns selected, they can't select all?

Reply:
(Same question for SelectTables).

Reply:
Yes, for the behavior, if they have selected some column, if they click on the toggle, then it will clear all selections. If they select again, then it will select all. Modeled the behavior from Gmail

Reply:
Ah, thanks for the extra context! It looks like Gmail has an indeterminate state of the checkbox for when some columns are selected but not all of them. I think it'd be helpful to do that here, too. Otherwise as a user, I might be surprised that checking the (empty) checkbox to select all unselects  them instead of selecting all.

Reply:
OK I see you already did this in a later PR! So ignore previous comment :)